src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncValueTaskMethodBuilderT.cs

```diff
@@ -460,10 +462,23 @@ private static void ExecutionContextCallback(object? s)
             public Action MoveNextAction => _moveNextAction ??= new Action(MoveNext);
 
             /// <summary>Invoked to run MoveNext when this instance is executed from the thread pool.</summary>
-            void IThreadPoolWorkItem.Execute() => MoveNext();
+            void IThreadPoolWorkItem.Execute()
+            {
+                ExecutionContext? context = Context;
+
+                if (context is null)
+                {
+                    Debug.Assert(!(StateMachine is null));
+                    StateMachine.MoveNext();
+                }
+                else
+                {
+                    ExecutionContext.RunForThreadPoolUnsafe(context, s_callback, this);
```

Comment:
We can use the lighterweight `RunForThreadPoolUnsafe` here; since its on the ThreadPool